scripts/db-heal-test.py

```diff
@@ -0,0 +1,231 @@
+import os
+import subprocess
+import tempfile
+import time
+import shutil
+from pathlib import Path
+import sys
+import flag
+
+# --- Dynamic variables --- #
+aida_path_flag = flag.string("aida", "", "Path to Aida root.")
+aida_db_path_flag = flag.string("aida_db", "", "Path to AidaDB.")
+tmp_path_flag = flag.string("tmp", "", "Path to tmp dir.")
+
+number_of_iterations_flag = flag.int("iter", 1000, "Number of iterations.")
+window_flag = flag.int("window", 120, "How often will program get terminated (in seconds).")
+checkpoint_granularity_flag = flag.int("cp_granularity", 2000, "How often will Carmen create checkpoints.")
+flag.parse()
+
+aida_path = aida_path_flag.value
+aida_db_path = aida_db_path_flag.value
+tmp_path = tmp_path_flag.value
+
+number_of_iterations = number_of_iterations_flag.value
+window = window_flag.value
+checkpoint_granularity = checkpoint_granularity_flag.value
+
+# Mark first checkpoint
+latest_checkpoint = checkpoint_granularity
+
+if aida_path == "":
+    print("please set Aida using --aida")
+    exit(1)
+if aida_db_path == "":
+    print("please set AidaDB using --aidadb")
+    exit(1)
+if tmp_path == "":
+    tmp_path = tempfile.gettempdir()
+    print(f"tmp not set - using default {tmp_path}")
+
+# Block variables
+first_block = 0
+last_block = 60000000
+
+# --- Static variables --- #
+
+carmen_root = os.path.abspath('../go')
+
+# --- Script --- #
+
+# Log file path from which we read output to find kill_block
+aida_log_file = Path.cwd() / 'aida.log'
+carmen_log_file = Path.cwd() / 'carmen.log'
+current_dir = Path.cwd()
+
+print("Your settings:")
+print(f"\tNumber of iterations: {number_of_iterations}.")
+print(f"\tWindow size: {window} seconds.")
+print(f"\tCheckpoint granularity: {checkpoint_granularity} blocks.")
+
+
+# Function which stops process after given sleep_time.
+def terminate_process_after(sleep_time: int, checkpoint: int):
```

Comment:
I think the last checkpoint and the time window when to kill the process should be tracked in two separate threads.

At the moment when I configure CP every 1M block and kill time 5s, the program is not interrupted every 5s - it is interrupted after each 1M blocks (close after this block heigh is done as the timeout is long overdue)

We wanted to test the 
1. a program is killed at arbitrary time  - i.e. may be a fixed window of X seconds
2. right when the program is killed, we check what is the last checkpoint and recover it

current solution IMHO still depends (waits for) a checkpoint being done and only after that it recovers. 

not sure if it is the best way to get the last checkpoint by parsing the log, ideally Carmen should tell what is the last CP, but not sure we have this option at the moment.

Reply:
Yes it does wait until the first CP is done, but thats the only way to actually know Aida started block processing. 
Then it waits some time chosen by `--window` and then it terminates Aidas process. I thought the aim of this script was to be able to terminate the program at the time checkpoint is being created, which does not mean it needs to kill it at the first checkpoint.

Also why would we even want to kill the process without having the checkpoint? 

With `--cp_granularity` you choose how often (in blocks) Carmen creates checkpoints. It also choses how often Aida logs progress, so that it logs the block number at the time when Carmen creates checkpoint (tldr `carmen-checkpoint-interval` == `aida-tracker-interval`)...
If you chose 1M blocks it makes sense the script wont kill the process in 5s because there is no way Aida could sync to 1M in that time window and I think this is correct behaviour so that you do not chose weird combination because the script would do nothing.

Reply:
The key message is that you do not kill at arbitrary time, but always wait after the CP is ready (and potentially then it wats a bit more).

if you want to measure time only after there is at least one CP, then yes, I agree it is a good idea - but it cannot break what we want to reach - kill at arbitrary (random) time